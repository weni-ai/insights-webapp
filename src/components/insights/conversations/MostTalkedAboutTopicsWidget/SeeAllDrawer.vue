<template>
  <UnnnicDrawer
    class="see-all-drawer"
    :modelValue="modelValue"
    :title="$t('conversations_dashboard.most_talked_about_topics.title')"
    size="lg"
    @close="emit('update:modelValue', false)"
  >
    <template #content>
      <ProgressTable
        :progressItems="MOCK_DATA_PROGRESS_TABLE"
        :subItemsDescription="
          $t('conversations_dashboard.most_talked_about_topics.subtopics')
        "
        :expandedItems="expandedItems"
      />
    </template>
  </UnnnicDrawer>
</template>

<script setup lang="ts">
import ProgressTable from '@/components/ProgressTable.vue';
import { addColors } from '@/utils/treemap';
import { useWidgetFormatting } from '@/composables/useWidgetFormatting';

const { formatPercentage, formatNumber } = useWidgetFormatting();

defineProps<{
  modelValue: boolean;
}>();

const emit = defineEmits<{
  (_e: 'update:modelValue', _value: boolean): void;
}>();

const MOCK_DATA = [
  {
    label: 'Entrega atrasada',
    value: 6973,
    percentage: 29,
    subItems: [
      {
        label: 'Atraso no transporte',
        value: 5229,
        percentage: 75,
      },
      {
        label: 'Problema na preparação',
        value: 1000,
        percentage: 25,
      },
    ],
  },
  {
    label: 'Produto defeituoso',
    value: 5500,
    percentage: 23,
  },
  {
    label: 'Dúvidas sobre preço',
    value: 1600,
    percentage: 16,
    subItems: [
      {
        label: 'Desconto no valor',
        value: 1825,
        percentage: 35,
      },
      {
        label: 'Desconto no frete',
        value: 375,
        percentage: 65,
      },
    ],
  },
  {
    label: 'Cancelamento',
    value: 1400,
    percentage: 14,
  },
  {
    label: 'Unclassified',
    value: 1000,
    percentage: 13,
  },
  {
    label: 'Outros',
    value: 500,
    percentage: 5,
  },
];

const coloredData = addColors(MOCK_DATA);

const MOCK_DATA_PROGRESS_TABLE = coloredData.map((item) => ({
  label: item.label,
  value: item.percentage,
  description: `${formatPercentage(item.percentage)} (${formatNumber(item.value)})`,
  backgroundColor: item.color,
  color: item.hoverColor,
  subItems: MOCK_DATA.find(
    (mockItem) => mockItem.label === item.label,
  )?.subItems?.map((subItem) => ({
    label: subItem.label,
    value: subItem.percentage,
    description: `${formatPercentage(subItem.percentage)} (${formatNumber(subItem.value)})`,
    backgroundColor: item.color,
    color: item.hoverColor,
  })),
}));

const expandedItems = defineModel<string[]>('expandedItems', {
  default: () => [],
});
</script>
