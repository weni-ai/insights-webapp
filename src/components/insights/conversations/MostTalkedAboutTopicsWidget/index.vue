<template>
  <section class="most-talked-about-topics" data-testid="most-talked-about-topics">
    <BaseConversationWidget
      class="most-talked-about-topics__conversation-widget"
      data-testid="topics-base-widget"
      :title="$t('conversations_dashboard.most_talked_about_topics.title')"
      :actions="[
        {
          icon: 'edit_square',
          text: $t(
            'conversations_dashboard.most_talked_about_topics.edit_topics_and_subtopics',
          ),
          onClick: toggleAddTopicsDrawer,
        },
      ]"
      @tab-change="handleTabChange"
    >
      <TreemapChart
        :isLoading="isLoadingTopicsDistribution"
        :data="treemapData"
        data-testid="topics-treemap-chart"
        @click="handleSeeAllDrawer($event.label)"
      />

      <UnnnicButton
        v-if="treemapData.length > 0"
        type="tertiary"
        size="small"
        :disabled="isLoadingTopicsDistribution"
        :text="$t('conversations_dashboard.most_talked_about_topics.see_all')"
        data-testid="topics-see-all-button"
        @click="handleSeeAllDrawer()"
      />

      <SeeAllDrawer
        v-if="isSeeAllDrawerOpen"
        v-model="isSeeAllDrawerOpen"
        :data="treemapData"
        :expandedItems="expandedItems"
        data-testid="topics-see-all-drawer"
      />
    </BaseConversationWidget>

    <AddWidget
      v-if="!hasExistingTopics && !isLoadingTopicsDistribution"
      :title="$t('conversations_dashboard.most_talked_about_topics.no_topics')"
      :description="
        $t(
          'conversations_dashboard.most_talked_about_topics.no_topics_description',
        )
      "
      :actionText="
        $t('conversations_dashboard.most_talked_about_topics.add_first_topic')
      "
      data-testid="topics-add-widget"
      @action="toggleAddTopicsDrawer"
    />

    <DrawerTopics data-testid="topics-drawer" />
  </section>
</template>

<script setup lang="ts">
import { computed, onMounted, ref, watch } from 'vue';
import { storeToRefs } from 'pinia';
import BaseConversationWidget from '@/components/insights/conversations/BaseConversationWidget.vue';
import TreemapChart from '@/components/insights/charts/TreemapChart.vue';
import SeeAllDrawer from './SeeAllDrawer.vue';
import AddWidget from '../AddWidget.vue';
import { useConversationalTopics } from '@/store/modules/conversational/topics';
import DrawerTopics from '../topics/DrawerTopics.vue';
import type { topicDistributionMetric } from '@/services/api/resources/conversational/topics';
import type { Tab } from '../BaseConversationWidget.vue';
import { useRoute } from 'vue-router';

const conversationalTopicsStore = useConversationalTopics();

const route = useRoute();

const MOCK_DATA: topicDistributionMetric[] = [
  {
    label: 'Entrega atrasada',
    value: 6973,
    percentage: 29,
    uuid: '',
  },
  {
    label: 'Produto defeituoso',
    value: 5500,
    percentage: 23,
    uuid: '',
  },
  {
    label: 'Dúvidas sobre preço',
    value: 1600,
    percentage: 16,
    uuid: '',
  },
  {
    label: 'Cancelamento',
    value: 1400,
    percentage: 14,
    uuid: '',
  },
  {
    label: 'Unclassified',
    value: 1000,
    percentage: 13,
    uuid: '',
  },
  {
    label: 'Outros',
    value: 500,
    percentage: 5,
    uuid: '',
  },
];

const {
  topicsDistributionCount,
  topicsDistribution,
  isLoadingTopicsDistribution,
  topicType,
  hasExistingTopics,
} = storeToRefs(conversationalTopicsStore);

const { loadTopicsDistribution, toggleAddTopicsDrawer, setTopicType } =
  conversationalTopicsStore;

const expandedItems = ref<string[]>([]);

const isSeeAllDrawerOpen = ref(false);
const handleSeeAllDrawer = (expandedItem?: string) => {
  if (isLoadingTopicsDistribution.value) return;

  expandedItems.value = expandedItem ? [expandedItem] : [];

  isSeeAllDrawerOpen.value = !isSeeAllDrawerOpen.value;
};

const treemapData = computed(() => {
  if (!hasExistingTopics.value) return MOCK_DATA;

  if (topicsDistributionCount.value === 0) return [];

  return topicsDistribution.value;
});

const handleTabChange = (tab: Tab) => {
  setTopicType(tab === 'artificial-intelligence' ? 'AI' : 'HUMAN');
};

watch(topicType, () => {
  loadTopicsDistribution();
});

watch(
  () => route.query,
  () => {
    loadTopicsDistribution();
  },
);

onMounted(() => {
  loadTopicsDistribution();
});
</script>

<style lang="scss" scoped>
.most-talked-about-topics {
  position: relative;

  padding-bottom: $unnnic-spacing-sm;

  overflow: hidden;

  min-width: 100%;
  height: 380px;

  &__conversation-widget {
    height: 100%;

    gap: $unnnic-spacing-ant;
  }
}
</style>
