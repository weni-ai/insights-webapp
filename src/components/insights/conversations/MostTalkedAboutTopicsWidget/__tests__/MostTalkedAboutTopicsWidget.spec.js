import { beforeEach, describe, expect, it, vi } from 'vitest';
import { config, mount } from '@vue/test-utils';
import { createI18n } from 'vue-i18n';

import MostTalkedAboutTopicsWidget from '../index.vue';

const mockTopicsStore = {
  topicsDistributionCount: { value: 5 },
  topicsDistribution: { value: [] },
  isLoadingTopicsDistribution: { value: false },
  topicType: { value: 'HUMAN' },
  hasExistingTopics: { value: true },
  loadTopicsDistribution: vi.fn(),
  toggleAddTopicsDrawer: vi.fn(),
  setTopicType: vi.fn(),
};

const mockRoute = {
  query: {},
};

vi.mock('@/store/modules/conversational/topics', () => ({
  useConversationalTopics: () => mockTopicsStore,
}));

vi.mock('vue-router', async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    useRoute: () => mockRoute,
  };
});

vi.mock('pinia', async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    storeToRefs: (store) => store,
  };
});

const i18n = createI18n({
  legacy: false,
  locale: 'en',
  messages: {
    en: {
      conversations_dashboard: {
        most_talked_about_topics: {
          title: 'Most Talked About Topics',
          edit_topics_and_subtopics: 'Edit Topics',
          see_all: 'See All',
          no_topics: 'No Topics',
          no_topics_description: 'No topics description',
          add_first_topic: 'Add First Topic',
        },
      },
    },
  },
  fallbackWarn: false,
  missingWarn: false,
});

config.global.plugins = [i18n];

describe('MostTalkedAboutTopicsWidget', () => {
  let wrapper;

  const createWrapper = (storeOverrides = {}) => {
    Object.assign(mockTopicsStore, storeOverrides);
    return mount(MostTalkedAboutTopicsWidget, {
      global: {
        stubs: {
          BaseConversationWidget: true,
          TreemapChart: true,
          UnnnicButton: true,
          SeeAllDrawer: true,
          AddWidget: true,
          DrawerTopics: true,
        },
      },
    });
  };

  const section = () =>
    wrapper.find('[data-testid="most-talked-about-topics"]');

  beforeEach(() => {
    vi.clearAllMocks();
    Object.assign(mockTopicsStore, {
      topicsDistributionCount: { value: 5 },
      topicsDistribution: { value: [] },
      isLoadingTopicsDistribution: { value: false },
      topicType: { value: 'HUMAN' },
      hasExistingTopics: { value: true },
    });
    wrapper = createWrapper();
  });

  describe('Component rendering', () => {
    it('should render main section', () => {
      expect(section().exists()).toBe(true);
    });

    it('should render with existing topics', () => {
      wrapper = createWrapper({
        hasExistingTopics: { value: true },
        isLoadingTopicsDistribution: { value: false },
      });

      expect(section().exists()).toBe(true);
    });

    it('should render when loading', () => {
      wrapper = createWrapper({
        hasExistingTopics: { value: false },
        isLoadingTopicsDistribution: { value: true },
      });

      expect(section().exists()).toBe(true);
    });
  });

  describe('Component structure', () => {
    it('should have correct CSS classes', () => {
      expect(section().classes()).toContain('most-talked-about-topics');
    });

    it('should match snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('Data management', () => {
    it('should use empty array when topics count is zero', () => {
      wrapper = createWrapper({
        hasExistingTopics: { value: true },
        topicsDistributionCount: { value: 0 },
      });

      expect(wrapper.vm.treemapData).toEqual([]);
    });

    it('should use store distribution when has existing topics', () => {
      const mockDistribution = [{ label: 'Topic', value: 100, percentage: 50 }];
      wrapper = createWrapper({
        hasExistingTopics: { value: true },
        topicsDistributionCount: { value: 1 },
        topicsDistribution: { value: mockDistribution },
      });

      expect(wrapper.vm.treemapData).toEqual(mockDistribution);
    });
  });

  describe('Event handling', () => {
    it('should call handleSeeAllDrawer method', () => {
      expect(wrapper.vm.isSeeAllDrawerOpen).toBe(false);

      wrapper.vm.handleSeeAllDrawer();

      expect(wrapper.vm.isSeeAllDrawerOpen).toBe(true);
      expect(wrapper.vm.expandedItems).toEqual([]);
    });

    it('should handle expanded item in handleSeeAllDrawer', () => {
      wrapper.vm.handleSeeAllDrawer('Specific Topic');

      expect(wrapper.vm.isSeeAllDrawerOpen).toBe(true);
      expect(wrapper.vm.expandedItems).toEqual(['Specific Topic']);
    });

    it('should not toggle drawer when loading', () => {
      wrapper = createWrapper({ isLoadingTopicsDistribution: { value: true } });

      wrapper.vm.handleSeeAllDrawer();

      expect(wrapper.vm.isSeeAllDrawerOpen).toBe(false);
    });

    it('should call setTopicType for AI tab', () => {
      wrapper.vm.handleTabChange('artificial-intelligence');

      expect(mockTopicsStore.setTopicType).toHaveBeenCalledWith('AI');
    });

    it('should call setTopicType for HUMAN tab', () => {
      wrapper.vm.handleTabChange('human');

      expect(mockTopicsStore.setTopicType).toHaveBeenCalledWith('HUMAN');
    });
  });

  describe('Lifecycle', () => {
    it('should load topics distribution on mounted', () => {
      expect(mockTopicsStore.loadTopicsDistribution).toHaveBeenCalled();
    });
  });

  describe('Conditional rendering', () => {
    it('should handle different store states', () => {
      const testCases = [
        {
          hasExistingTopics: { value: true },
          isLoadingTopicsDistribution: { value: false },
        },
        {
          hasExistingTopics: { value: false },
          isLoadingTopicsDistribution: { value: false },
        },
        {
          hasExistingTopics: { value: false },
          isLoadingTopicsDistribution: { value: true },
        },
      ];

      testCases.forEach((testCase) => {
        wrapper = createWrapper(testCase);
        expect(section().exists()).toBe(true);
      });
    });
  });
});
