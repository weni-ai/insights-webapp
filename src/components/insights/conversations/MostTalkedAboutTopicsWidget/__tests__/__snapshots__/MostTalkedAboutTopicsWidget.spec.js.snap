// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MostTalkedAboutTopicsWidget > Component structure > should match snapshot 1`] = `
<section
  class="most-talked-about-topics"
  data-testid="most-talked-about-topics"
  data-v-56fece4b=""
>
  <base-conversation-widget-stub
    actions="[object Object]"
    class="most-talked-about-topics__conversation-widget"
    data-testid="topics-base-widget"
    data-v-56fece4b=""
    hiddentabs="false"
    isloading="false"
    isonlytab="false"
    title="conversations_dashboard.most_talked_about_topics.title"
  />
  <!--v-if-->
  <drawer-topics-stub
    data-testid="topics-drawer"
    data-v-56fece4b=""
  />
</section>
`;
